File: go/algorithms/1.two-sum.go
package algorithms

func twoSum(nums []int, target int) []int {

	// Brute Force
	// for i := 0; i < len(nums); i++ {
	// 	for j := i + 1; j < len(nums); j++ {
	// 		if nums[i]+nums[j] == target {
	// 			return []int{i, j}
	// 		}
	// 	}
	// }

	// Hashmap to store the number as a key, and store the index as a value
	m := make(map[int]int)

	// Loop through the array
	for i, num := range nums {

		// Check if the corresponding number is already in the map or not.
		if index, ok := m[target-num]; ok {

			// Return the index of corresponding number and the index of the current number
			return []int{index, i}
		}
		// Store the number and the index in the map
		m[num] = i
	}

	return nil
}

File: python/hash-table/1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#
# https://leetcode.com/problems/two-sum/description/
#
# algorithms
# Easy (54.51%)
# Likes:    59469
# Dislikes: 2127
# Total Accepted:    15.6M
# Total Submissions: 28.6M
# Testcase Example:  '[2,7,11,15]\n9'
#
# Given an array of integers nums and an integer target, return indices of the
# two numbers such that they add up to target.
#
# You may assume that each input would have exactly one solution, and you may
# not use the same element twice.
#
# You can return the answer in any order.
#
#
# Example 1:
#
#
# Input: nums = [2,7,11,15], target = 9
# Output: [0,1]
# Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
#
#
# Example 2:
#
#
# Input: nums = [3,2,4], target = 6
# Output: [1,2]
#
#
# Example 3:
#
#
# Input: nums = [3,3], target = 6
# Output: [0,1]
#
#
#
# Constraints:
#
#
# 2 <= nums.length <= 10^4
# -10^9 <= nums[i] <= 10^9
# -10^9 <= target <= 10^9
# Only one valid answer exists.
#
#
#
# Follow-up: Can you come up with an algorithm that is less than O(n^2) time
# complexity?
#

from typing import List


# @lc code=start
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Hashmap to store the number as a key, and store the index as a value
        hash_map = {}
        for i, num in enumerate(nums):
            # Check if the corresponding number is already in the map or not.
            if target - num in hash_map:
                # Return the index of corresponding number
                # and the index of the current number
                return [hash_map[target - num], i]
            # Store the number and the index in the map
            hash_map[num] = i

        # Brute force solution
        # for i in range(len(nums)):
        #     for j in range(i+1, len(nums)):
        #         if nums[i] + nums[j] == target:
        #             return [i, j]

        return []


# @lc code=end
